### PR TITLE
Add CVaR scenario tail excess constraints

### LINKED ISSUE #1528: [CVaR]: Add tail excess constraint
### Description

- Add constraint for tail excess for s ∈ S. Use descriptive name `tail_excess`
  - Add a new file for this constraint here: https://github.com/TulipaEnergy/TulipaEnergyModel.jl/tree/main/src/constraints
  - Add the corresponding SQL query: https://github.com/TulipaEnergy/TulipaEnergyModel.jl/blob/main/src/sql/create-constraints.sql
  - Only create the new constraints if the value of the $\lambda$ parameter (see #1524) is greater than zero
- It depends on the scenario index, and the units are the currency of the objective function
- Add a test that checks that the constraints are created when the $\lambda$ parameter is greater than zero (test for one and for more than one scenarios)

### PR DESCRIPTION
**Pull request by datejada**
The main changes include adding new expressions and constraints for scenario tail excess, integrating these into the model creation workflow, and refactoring some objective function components for clarity and consistency.

The changes read better by commit: 

- [Update obj. func expression registration](https://github.com/TulipaEnergy/TulipaEnergyModel.jl/commit/a3ba5f498beb003ac71a1250171a007efcb90dfb)

- [Enforce scenario ordering in CVaR variable and constraint tables](https://github.com/TulipaEnergy/TulipaEnergyModel.jl/commit/fb3d09f81ddbbb70766e81438b4ec07face81043) 

- [Extract per-scenario operational cost expressions from objective](https://github.com/TulipaEnergy/TulipaEnergyModel.jl/commit/eab9a64e08f2fc159b3c25f4dca46a93f9239ce4) 

- [Add CVaR tail excess expressions and constraints](https://github.com/TulipaEnergy/TulipaEnergyModel.jl/commit/dd6485847aefddec31f1ef07db9e3a6a35f75c3b) 

Co-authored-by: GitHub Copilot <github-copilot[bot]@users.noreply.github.com>

## Related issues

Closes #1528 

## Checklist

- [x] I am following the [contributing guidelines](https://github.com/TulipaEnergy/TulipaEnergyModel.jl/blob/main/docs/src/90-contributing/90-contributing.md)
- [x] Tests are passing
- [x] Lint workflow is passing
- [x] Docs were updated and workflow is passing

### CODE REVIEW

**Comment by datejada** [comment]
@abelsiqueira, this PR is a bit big, so I would appreciate your review. I co-developed it using Copilot and following the AGENTS.md. After a couple of iterations, I am happy with this version for your review.

EDIT: ~I still want to add a constraint test using TulipaBuilder, but in the meantime, the current tests cover the basic ones for that constraint. So, you can start the review while I work on it.~ Done ✅ 

Thanks!

P.S. the Julia MCP is amazing 🤓

**Comment by datejada** [comment]
@abelsiqueira I was expecting the PR to be slower with these changes in the benchmarks, but I think the slowdown is not significant. What do you think?

File: src/objectives/create.jl [hunk 1]
```diff
@@ -1,15 +1,13 @@
...
+    _add_to_objective!(connection, objective_expr, name, expr)
 
-Add `expr` to the running objective sum, register it on the model under `name`, and
-insert a placeholder row in the `obj_breakdown` table (value filled in by
-`save_solution!` after the solve).
+Add `expr` to the running objective sum  and insert a placeholder row
+in the `obj_breakdown` table (value filled in by `save_solution!` after the solve).
 """
-function _add_to_objective!(connection, model, objective_expr, name::String, expr)
+function _add_to_objective!(connection, objective_expr, name::String, expr)
```
Comment (abelsiqueira) [inline src/objectives/create.jl:7]: I'm reading the commits one at a time. I don't know why this commit was made, i.e., why remove `model` from `_add_to_objective!` and why change how the expressions were defined

File: src/sql/create-variables.sql [hunk 2]
```diff
@@ -539,13 +539,22 @@ drop table if exists var_tail_excess_slack_xi
...
 create table var_tail_excess_slack_xi as
+with ordered_scenarios as (
+    select
+        scenario,
+        probability
+    from
+        stochastic_scenario
+    order by
+        scenario
+)
```
Comment (abelsiqueira) [inline src/sql/create-variables.sql:588]: I'm not really opposed, just wondering if you have a reason to order by scenario

File: test/test-objective-operational-costs-per-scenario.jl [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        expr_rows = DuckDB.query(
+            connection,
+            """
+            SELECT
+                scenario,
+                probability
+            FROM $(expr.table_name)
+            ORDER BY id
+            """,
+        ) |> collect
```
Comment (abelsiqueira) [inline test/test-objective-operational-costs-per-scenario.jl]: This could be a DataFrame (it's already loaded in the tests). It would make the tests easier to follow

File: test/test-objective-operational-costs-per-scenario.jl [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    scenario_rows = DuckDB.query(
+        connection,
+        """
+        SELECT
+            scenario,
+            probability
+        FROM stochastic_scenario
+        ORDER BY scenario
+        """,
+    ) |> collect
```
Comment (abelsiqueira) [inline test/test-objective-operational-costs-per-scenario.jl]: As well

File: src/expressions/operational-costs.jl [hunk 5]
```diff
@@ -0,0 +1,375 @@
...
+        return Dict{Int32,UnitRange{Int}}()
+    end
+
+    scenario_ids = [row.scenario for row in indices]
+    group_starts = [1; [i for i in 2:n if scenario_ids[i] != scenario_ids[i-1]]]
+    group_ends = [group_starts[2:end] .- 1; n]
+    scenarios = @view scenario_ids[group_starts]
+
+    return Dict(scenarios[k] => group_starts[k]:group_ends[k] for k in eachindex(scenarios))
+end
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl]: This needs a docstring explaining what it does and why we need it

File: src/expressions/operational-costs.jl [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+            )
+            costs[i] =
+                row.cost_coefficient::Float64 *
+                (
+                    row.commodity_price::Float64 * commodity_price_agg /
+                    row.producer_efficiency::Float64 + row.operational_cost::Float64
+                ) *
+                (row.time_block_end - row.time_block_start + 1)
+        end
+    end
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl]: what's happening here?

File: src/expressions/operational-costs.jl (hunk 5 above)
Comment (abelsiqueira) [reply src/expressions/operational-costs.jl]: I'll have to recheck this section to understand it after the docstring and the other comment above

File: src/constraints/conditional_value_at_risk.jl [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+)
+    let table_name = :scenario_tail_excess, cons = constraints[table_name]
+        indices = _append_scenario_tail_excess_data_to_indices(connection, table_name)
+        if isempty(variables[:tail_excess_slack_xi].container)
+            return nothing
+        end
+        var_tail_excess_slack_xi = variables[:tail_excess_slack_xi].container
+        total_cost_per_scenario =
+            expressions[:scenario_tail_excess].expressions[:total_cost_per_scenario]
+        var_value_at_risk_threshold_mu = variables[:value_at_risk_threshold_mu].container[1]
```
Comment (abelsiqueira) [inline src/constraints/conditional_value_at_risk.jl:36]: Use `only` instead to be clear that it should be only 1

File: src/expressions/conditional_value_at_risk.jl [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        attach_expression!(
+            expressions[:scenario_tail_excess],
+            :total_cost_per_scenario,
+            JuMP.AffExpr[],
+        )
+        return nothing
+    end
+
+    let table_name = :scenario_tail_excess, expr = expressions[table_name]
+        n = expr.num_rows
```
Comment (abelsiqueira) [inline src/expressions/conditional_value_at_risk.jl:32]: I'd rather be explicit below

File: test/test-constraint-scenario-tail-excess.jl [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    DuckDB.query(
+        connection,
+        """
+        CREATE OR REPLACE TABLE model_parameters AS
+        SELECT *
+        FROM (VALUES
+        	(0.1, 0.98)
+        ) AS t(risk_aversion_weight_lambda, risk_aversion_confidence_level_alpha);
+        """,
+    )
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl]: There is a function to create tables in the tests, I think. If so, also ask AI to update AGENTS with that info

File: test/test-constraint-scenario-tail-excess.jl [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    energy_problem = TulipaEnergyModel.EnergyProblem(connection)
+    TulipaEnergyModel.create_model!(energy_problem)
+
+    total_cost_per_scenario =
+        energy_problem.expressions[:scenario_tail_excess].expressions[:total_cost_per_scenario]
+
+    model = energy_problem.model
+    tail_excess_indices = energy_problem.variables[:tail_excess_slack_xi].indices
+    tail_excess_vars = energy_problem.variables[:tail_excess_slack_xi].container
+    value_at_risk_threshold_mu = energy_problem.variables[:value_at_risk_threshold_mu].container[1]
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl]: only

File: test/test-constraint-scenario-tail-excess.jl [hunk 11]
```diff
@@ -0,0 +1,112 @@
+@testitem "Create scenario tail excess for case study" setup = [CommonSetup] tags =
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl]: The name feels off, but if you like it, that's fine

File: test/test-constraint-scenario-tail-excess.jl [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+            expected_total_cost,
+            units_on_operational_cost_per_scenario[row.scenario],
+        )
+
+        @test expected_total_cost == total_cost_per_scenario[row.id]
+
+        expected_cons = JuMP.@build_constraint(
+            tail_excess_vars[row.id] >=
+            total_cost_per_scenario[row.id] - value_at_risk_threshold_mu
+        )
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl:255]: It's hard to follow which constraint is expected to be implemented.
Maybe we can be explicit here with the values?

File: test/test-constraint-scenario-tail-excess.jl [hunk 13]
```diff
@@ -1,3 +1,186 @@
...
+        ((investment_cost, assets_investment), (decommission_cost, assets_decommission))
+        JuMP.add_to_expression!(base_cost, coef, var)
+    end
+
+    # Create expected flows operational cost per scenario
+    operational_cost_per_scenario = Dict{Int64,JuMP.AffExpr}()
+    for scenario in 1:num_scenarios
+        operational_cost_per_scenario[scenario] = JuMP.AffExpr(0.0)
+        for rp in 1:num_rep_periods
+            weight = get_weight(connection, rp, scenario)
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl]: It should be faster to create `weights_lookup` by

```julia
weight_lookup = Dict((row.rp, row.scenario) => row.weight) for row in DuckDB.query(...))
```

Can still be split into a separate function call if necessary.

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Hi @datejada, thanks for the PR.

This was a bit hard to follow, and it's possible some things were left behind. In particular, there was code moving between files that I assume was correct.
I left a few comments per commit. Sometimes the suggestion function didn't work.

File: test/test-constraint-scenario-tail-excess.jl (hunk 13 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl]: Great! I like it better. Changed in the latest update.

File: test/test-constraint-scenario-tail-excess.jl (hunk 12 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl:255]: There is an update to make more clear (I hope). Lines https://github.com/TulipaEnergy/TulipaEnergyModel.jl/pull/1575/changes#diff-87fdb66609efbeb19cb26ed96255336168cfc850de9876f31c30e0532edd0632R137-R160

Construct the terms from the input data; I think it reads better from there what you expect. The tricky part is to create it per scenario.

I can write term by term, instead of creating the expressions, to make it more explicit. Let me know if the change improved or if you prefer the explicit version. Both are okay.

Thanks

File: test/test-constraint-scenario-tail-excess.jl (hunk 11 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl]: True, I changed it so it read clearly

File: test/test-constraint-scenario-tail-excess.jl (hunk 10 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl]: updated here and in all the places that I have used it for testing.

File: test/test-constraint-scenario-tail-excess.jl (hunk 9 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl]: True, I forgot about that one. It is updated and also added to the AGENTS.md

File: src/expressions/conditional_value_at_risk.jl (hunk 8 above)
Comment (datejada) [reply src/expressions/conditional_value_at_risk.jl:32]: I am not sure if I follow the comment. Do you mean, not using `Let`? I was trying to keep the same structure as in other constraints. But elaborate a bit more on your comment; maybe it's something else you are referring to.

File: src/constraints/conditional_value_at_risk.jl (hunk 7 above)
Comment (datejada) [reply src/constraints/conditional_value_at_risk.jl:36]: done!

File: src/expressions/operational-costs.jl (hunk 6 above)
Comment (datejada) [reply src/expressions/operational-costs.jl]: The main structure is the same as in main:
https://github.com/TulipaEnergy/TulipaEnergyModel.jl/blob/f95cb9193fa11ea545f4d6862786bf858a17dee2/src/objectives/flows_operational_cost.jl#L96-L117

But here we need the cost per scenario rather than the total costs across scenarios. I have added docstrings to the functions to help follow what is happening in this function. I hope it helps.

Let me know if it is enough or if you want more details.

File: src/expressions/operational-costs.jl (hunk 5 above)
Comment (datejada) [reply src/expressions/operational-costs.jl]: I have added docstrings to all the functions, including this one to help understand what is happening here.

I created this function with the help of Copilot. Here is the explanation of why this is a good implementation for the purpose of this PR (I think it is a good approach, but please let me know if you think differently):

```text
Why _scenario_ranges is needed
The three SQL queries (_query_flows_operational_cost_per_scenario_indices, etc.) return flat rows sorted by (scenario, var.id). To build a Vector{JuMP.AffExpr} with one expression per scenario, the code needs to group those rows by scenario. _scenario_ranges does exactly that in O(n) — detecting contiguous run boundaries — and returns a Dict{Int32, UnitRange{Int}} used to slice the flat array per scenario.

Without it, each of the 3 expression builders would need either:

Additional SQL queries (one per scenario), or
An O(n × S) scan (filter the full array for each scenario)
It's called 3 times (once per expression type) and its output directly enables the CVaR expression builder in conditional_value_at_risk.jl to access expressions[:flows_operational_cost_per_scenario].expressions[:cost][i] by position.

The current _scenario_ranges approach is more efficient. Here's why:

SQL-per-scenario approach

S separate DuckDB queries, each with: query parsing, planning, execution, and result materialization overhead
DuckDB round-trip cost is non-trivial (milliseconds per call)
Total cost scales as O(S × query_overhead + n)
Current _scenario_ranges approach

1 SQL query fetching all n rows, already sorted by (scenario, var.id)
1 O(n) Julia pass to detect contiguous run boundaries
Total cost: O(query_overhead + n)
The SQL results are already sorted by scenario (the ORDER BY rp_weight.scenario, var.id at the end of each query). _scenario_ranges exploits that sort to do a single linear scan — it just detects where consecutive scenario_ids differ. This is essentially free compared to S additional query round-trips.

In stochastic optimization problems the number of scenarios S can be large (tens to hundreds), making S repeated queries significantly more expensive than the current O(n) scan. The current implementation is the right choice.
```

File: test/test-objective-operational-costs-per-scenario.jl (hunk 4 above)
Comment (datejada) [reply test/test-objective-operational-costs-per-scenario.jl]: Updated, thanks!

File: test/test-objective-operational-costs-per-scenario.jl (hunk 3 above)
Comment (datejada) [reply test/test-objective-operational-costs-per-scenario.jl]: Good point. It was updated. Thanks

File: src/sql/create-variables.sql (hunk 2 above)
Comment (datejada) [reply src/sql/create-variables.sql:588]: This is to ensure the alignment when building the constraints. I got an error when running one example because I gave the scenario order 1998, 2009, 2008. So, since the assumption of order is other constraints, I followed that idea and ensured the order here as well for this feature.

File: src/objectives/create.jl (hunk 1 above)
Comment (datejada) [reply src/objectives/create.jl:7]: The main reason is that I want to store in an expression the total cost per component and scenario for this new constraint. But, in the objective function, the cost is multiplied by a parameter (1-lambda), so if registered here, then it includes the (1-lambda) factor. So, that is why I register the expression first, before calling the _add_to_objective! without the (1-lamdba) (to reuse in the constraint) but I pass to the objective the expression times the (1-lamdba). This follows the formulation in here:

https://github.com/TulipaEnergy/TulipaEnergyModel.jl/issues/1523#issuecomment-4044693505

I think with that context, the change makes more sense, but please let me know if you need more details. Thanks!

**Comment by datejada** [comment]
> Hi @datejada, thanks for the PR.
> 
> This was a bit hard to follow, and it's possible some things were left behind. In particular, there was code moving between files that I assume was correct. I left a few comments per commit. Sometimes the suggestion function didn't work.

Thanks, @abelsiqueira. Sorry, this is a huge and not easy-to-follow PR. Also, I didn't provide enough context for the PR's aim. It boils down to trying to implement constraint (6j) here:

https://github.com/TulipaEnergy/TulipaEnergyModel.jl/issues/1523#issuecomment-4044693505

For that, you would need the cost per scenario. So, I tried to create expressions that I was able to reuse in the constraint. Otherwise, I would be repeating the cost calculation in two places. In addition, the cost in the objective function is multiplied by (1-lambda). So, the expressions don't include that, so I can reuse them in the constraint.

Anyway, I have implemented the changes and replied to your comments. Please let me know if you have more comments or suggestions.

Thanks!

File: src/sql/create-variables.sql (hunk 2 above)
Comment (abelsiqueira) [reply src/sql/create-variables.sql:588]: Makes sense. I'll tag #1335 here to create the x-ref in case we work on it in the future
Marking as resolved

File: src/expressions/operational-costs.jl (hunk 6 above)
Comment (abelsiqueira) [reply src/expressions/operational-costs.jl]: Got it. I think the `collect` might become a problem, especially for larger results.
So something like

```julia
costs = [
    ...
    for row in indices
]
```

might be necessary, like we do in the expressions and constraints

File: src/expressions/operational-costs.jl [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+    connection,
+    model,
+    variables,
+    expressions,
+)
+    expr_name = :vintage_flows_operational_cost_per_scenario
+    expr = _create_scenario_cost_expression!(connection, expressions, expr_name)
+    expr_indices = DuckDB.query(connection, "FROM expr_$expr_name ORDER BY id") |> collect
+
+    indices = _query_vintage_flows_operational_cost_per_scenario_indices(connection) |> collect
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl]: Here again we're collecting. It's more complicated to fix this one, since you want to relate two tables.
Our normal strategy would be to use `array_agg` to link one index to another, but here we have `_scenario_ranges` building ranges, so I don't know what is possible.
Is the benchmark picking up this change? I would expect some slowdown here if it is. If there is no slowdown even with the benchmark picking it up, then I'm overthinking it.

File: src/expressions/conditional_value_at_risk.jl (hunk 8 above)
Comment (abelsiqueira) [reply src/expressions/conditional_value_at_risk.jl:32]: Sorry, I meant to use `expr.num_rows` explicitly below instead of `n`

File: src/expressions/conditional_value_at_risk.jl [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+        flows_operational_cost_per_scenario =
+            expressions[:flows_operational_cost_per_scenario].expressions[:cost]
+        vintage_flows_operational_cost_per_scenario =
+            expressions[:vintage_flows_operational_cost_per_scenario].expressions[:cost]
+        units_on_operational_cost_per_scenario =
+            expressions[:units_on_operational_cost_per_scenario].expressions[:cost]
+
+        # Construct the total cost per scenario expression for each scenario
+        total_cost_per_scenario = Vector{JuMP.AffExpr}(undef, n)
+        for i in 1:n
```
Comment (abelsiqueira) [inline src/expressions/conditional_value_at_risk.jl]:
```suggestion
        total_cost_per_scenario = Vector{JuMP.AffExpr}(undef, expr.num_rows)
        for i in 1:expr.num_rows
```

File: src/expressions/operational-costs.jl (hunk 14 above)
Comment (abelsiqueira) [reply src/expressions/operational-costs.jl]: Looking around, it looks like there `_query` functions are only used once each. Their main goal being to create these indices, which are then aggregated below per scenario to create an expression.
If that's the case, then it might be possible to create `group by scenario` and use `array_agg(id) as var_vintage_flow_ids, array_agg(cost) as arr_costs` and loop over those instead in the `attach_expression!`.
This is the same idea as the group investment in `src/constraints/investment_group.jl`.
Sorry if this is obviously not the case, but I can't see all the connected pieces clearly.

File: src/expressions/operational-costs.jl [hunk 16]
```diff
@@ -0,0 +1,473 @@
...
+                    sum(
+                        indices[i].cost * vintage_flow[indices[i].id::Int64] for
+                        i in range_per_scenario[row.scenario]
+                    ),
+                )
+            else
+                @expression(model, 0.0)
+            end for row in expr_indices
+        ],
+    )
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl:171]: Code suggestion related to the discussion above.
Assuming it's possible to group the stuff from `indices` and join into `expr_indices`

```suggestion
    attach_expression!(
        expr,
        :cost,
        JuMP.AffExpr[
            if length(row.var_vintage_flow_ids) > 0
                @expression(
                    model,
                    sum(
                        cost * vintage_flow[var_id::Int64]
                        for (cost, var_id) in zip(row.arr_cost, row.var_vintage_flow_ids)
                    ),
                )
            else
                @expression(model, 0.0)
            end for row in expr_indices
        ],
    )
```

File: src/expressions/operational-costs.jl (hunk 5 above)
Comment (abelsiqueira) [reply src/expressions/operational-costs.jl]: I made comments in other places. The AI answer is too localized. It justifies this function assuming the other decisions around the function are also necessary. I made a few comments in a few places that maybe clarify what I'm essence, this raised a flag because in my mind it looks like an aggregation of the indices, which we do directly in SQL in other places

File: test/test-constraint-scenario-tail-excess.jl [hunk 17]
```diff
@@ -0,0 +1,279 @@
...
+        cost_per_scenario = JuMP.AffExpr(0.0)
+        JuMP.add_to_expression!(cost_per_scenario, base_cost)
+        JuMP.add_to_expression!(cost_per_scenario, flows_operational_cost_per_scenario[scenario])
+        JuMP.add_to_expression!(
+            cost_per_scenario,
+            vintage_flows_operational_cost_per_scenario[scenario],
+        )
+        JuMP.add_to_expression!(cost_per_scenario, units_on_operational_cost_per_scenario[scenario])
+
+        @test cost_per_scenario == total_cost_per_scenario[id]
```
Comment (abelsiqueira) [inline test/test-constraint-scenario-tail-excess.jl]:
```suggestion
        @test cost_per_scenario ≈ total_cost_per_scenario[id]
```

File: test/test-constraint-scenario-tail-excess.jl (hunk 12 above)
Comment (abelsiqueira) [reply test/test-constraint-scenario-tail-excess.jl:255]: That looks good enough, I think, thanks

**Review by abelsiqueira — CHANGES_REQUESTED** [review]
Thanks, @datejada. I've made a few more comments now that I get the picture better

File: test/test-constraint-scenario-tail-excess.jl (hunk 17 above)
Comment (datejada) [reply test/test-constraint-scenario-tail-excess.jl]: Actually, it is not possible to do it like this. It throws an error: "approx doesn't exist" when comparing JuMP expressions. This test compares the expression, not its value. So, `==` is correct here. If we were solving the problem, then comparing the values with approx would be correct.

File: src/expressions/operational-costs.jl (hunk 5 above)
Comment (datejada) [reply src/expressions/operational-costs.jl]: I see your point and I agree, I didn't think about it like that, but it is true, it is an aggregation at the end. So, having consistency with other parts of the code is a good idea. I will change it to use SQL aggregations. Thanks!

**Comment by datejada** [comment]
@abelsiqueira, thank you for your comments. I have updated the code, and I think now it is more aligned with the other standards we have elsewhere, as you suggested. Here is a summary of the main changes (I also had to update the tests when rebasing):

**1. Replace `collect` + Julia-side grouping with SQL `ARRAY_AGG`**

All aggregation was moved into SQL using `ARRAY_AGG(... ORDER BY id)` grouped by scenario, so each query returns **one row per scenario** carrying all the per-flow/per-vintage-flow/per-unit arrays.

**2. Delete `_scenario_ranges`**

The helper function `_scenario_ranges` (which computed per-scenario index ranges from a flat sorted result set) became unused after the SQL `ARRAY_AGG` refactor and was removed.

Please let me know if you have any further comment or suggestion

File: src/expressions/operational-costs.jl [hunk 18]
```diff
@@ -70,53 +71,53 @@ function _add_flows_operational_cost_per_scenario_expressions!(
...
-        end
-    end
 
-    range_per_scenario = _scenario_ranges(indices)
+    indices = _query_flows_operational_cost_per_scenario_indices(
+        connection,
+        expr_name,
+        has_commodity_price_profile,
+    )
+
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl:82]: A `costs` array is being created inside the loop, so pre-allocating should be better, since we don't use the full vector, only arrays.
Something like

```julia
max_n_flows = get_single...("SELECT MAX(ARRAY_LENGTH(var_flow_ids)) FROM ...")
costs = Vector{Float64}(undef, max_n_flows)
```

File: src/expressions/operational-costs.jl [hunk 19]
```diff
@@ -70,53 +71,53 @@ function _add_flows_operational_cost_per_scenario_expressions!(
...
     attach_expression!(
         expr,
         :cost,
         JuMP.AffExpr[
-            if haskey(range_per_scenario, row.scenario)
+            if length(row.var_flow_ids) > 0
+                n_flows = length(row.var_flow_ids)
+                costs = Vector{Float64}(undef, n_flows)
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl]: Then this line can be removed

File: src/expressions/operational-costs.jl [hunk 20]
```diff
@@ -70,53 +71,53 @@ function _add_flows_operational_cost_per_scenario_expressions!(
...
+                                1.0,
+                            )
+                            row.arr_cost_coefficient[i]::Float64 *
+                            (
+                                row.arr_commodity_price[i]::Float64 * commodity_price_agg /
+                                row.arr_producer_efficiency[i]::Float64 +
+                                row.arr_operational_cost[i]::Float64
+                            ) *
+                            (row.arr_time_block_end[i] - row.arr_time_block_start[i] + 1)
+                        end
```
Comment (abelsiqueira) [inline src/expressions/operational-costs.jl]: To increase readability, this can be

```julia
if ...
    costs[i] = ...
else
    commodity_...
    costs[i] = ...
end
```

**Review by abelsiqueira — APPROVED** [review]
Hi @datejada, thanks for the update. It's very helpful. I've made three extra comments below, which are not blocking for approval, so it's pre-approved.

Two are about making a small modification to pre-allocate the costs. It shouldn't be hard to implement, but if it is, that's fine, we can skip it.
The last is about readability, but it's opinion.

I've made the comments in commits, so they're marked as outdated, I hope you can see them.

File: src/expressions/operational-costs.jl (hunk 18 above)
Comment (datejada) [reply src/expressions/operational-costs.jl:82]: Thanks! `costs` is now pre-allocated once before `attach_expression!` using `max_n_flows` (a single `SELECT MAX(cnt)` query over flows grouped by scenario).

File: src/expressions/operational-costs.jl (hunk 19 above)
Comment (datejada) [reply src/expressions/operational-costs.jl]: Great! With the change in the other comment, the `costs = Vector{Float64}(undef, n_flows)` allocation that was inside the per-scenario loop body is removed.

File: src/expressions/operational-costs.jl (hunk 20 above)
Comment (datejada) [reply src/expressions/operational-costs.jl]: I agree, it helps with the readability. So, the `costs[i] = if ... else ... end` form is replaced with the more readable `if ... costs[i] = ... else ... commodity_price_agg = ...; costs[i] = ... end` form.

**Comment by datejada** [comment]
> Hi @datejada, thanks for the update. It's very helpful. I've made three extra comments below, which are not blocking for approval, so it's pre-approved.
> 
> Two are about making a small modification to pre-allocate the costs. It shouldn't be hard to implement, but if it is, that's fine, we can skip it. The last is about readability, but it's opinion.
> 
> I've made the comments in commits, so they're marked as outdated, I hope you can see them.

Thanks @abelsiqueira, thanks for the suggestions. I have implemented all three of them 🤓
